src/transformers/trainer.py

```diff
@@ -1875,6 +1875,34 @@ def _inner_training_loop(
         self._memory_tracker.stop_and_update_metrics(metrics)
 
         self.log(metrics)
+        checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+
+        if self.hp_search_backend is not None and trial is not None:
+            if self.hp_search_backend == HPSearchBackend.OPTUNA:
+                run_id = trial.number
+            elif self.hp_search_backend == HPSearchBackend.RAY:
+                from ray import tune
+
+                run_id = tune.get_trial_id()
+            elif self.hp_search_backend == HPSearchBackend.SIGOPT:
+                run_id = trial.id
+            elif self.hp_search_backend == HPSearchBackend.WANDB:
+                import wandb
+
+                run_id = wandb.run.id
+            run_name = self.hp_name(trial) if self.hp_name is not None else f"run-{run_id}"
+            run_dir = os.path.join(self.args.output_dir, run_name)
+        else:
+            run_dir = self.args.output_dir
```

Comment:
Yes, this whole block could probably be refactored in a method `_get_output_dir(self, trial)`

Reply:
Done , But at #L2158 there is "self.store_flos()" inside the else part, not show how to handle it .

Reply:
You should:
1. use the `_get_output_dir` method
2. then make a test for `if self.hp_search_backend is None or trial is None:` and in it do the store_flos part.